# Соревнование для учащися курса "Специалист по работе с данными / Data Scientist". Школа 21 СБЕР / Томский государственный университет

https://www.kaggle.com/competitions/sber-21-tgu-2022/data

Задача - предсказать стоимость аренды квартир в Лондоне. Метрика - MAE.

Предварительный анализ данных см. в ноутбуке EDA.ipynb

In [1]:
import pandas as pd
import numpy as np
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_absolute_error
import chime

In [2]:
%load_ext chime

### Пропущенные значения

In [3]:
class Preprocess:
    def __init__(self, max_categories=6):
        self.max_categories = max_categories
        self.uncorrelated_features = ['latitude', 'longitude',
                                          'available', 'comments',
                                          'canceled', 'notice',
                                          'host_response_rate', 'host_since']
        self.correlated_features = ['accommodates', 'bathrooms', 'beds','guests_included',
                                        'security_deposit','bedrooms','minimum_nights',
                                        'extra_people','square_feet','cleaning_fee']

    def fit(self, X, y=None):    
        # Merge with other files
        calendar = pd.read_csv('calendar_availability.csv', index_col='listing_id')
        reviews = pd.read_csv('reviews_stat.csv', index_col='listing_id')
        df = X.copy().join([calendar, reviews], how='left')
        df.drop(columns='host_id', inplace=True)

        #Convert the column with the percent sign
        df['host_response_rate'] = df['host_response_rate']\
                .str.replace(r'\D+', '', regex=True)\
                .astype('float')

        # Convert datetimes into numbers for further use as a numeric variable
        # This will apply to the host_since column
        self.date_features = df.select_dtypes(include='datetime').columns.to_list()
        df[self.date_features] = df[self.date_features].astype(np.int64)

        self.numeric_features = df.select_dtypes(include='number').columns.to_list()
        self.cat_features = df.select_dtypes(include=['object', 'bool']).columns.to_list()
        

        self.scale_encode = ColumnTransformer(
            transformers = [
                ('scale', StandardScaler(), self.numeric_features),
                ('get_dummies', OneHotEncoder(sparse=False,
                                              drop='first', 
                                              handle_unknown='infrequent_if_exist',
                                              min_frequency=2,
                                             max_categories=self.max_categories), self.cat_features)
            ],
            remainder='passthrough',
            n_jobs=-1
        ).fit(df)
        
        return self

    
    def transform(self, X, y=None):
        # Merge with other files
        calendar = pd.read_csv('calendar_availability.csv', index_col='listing_id')
        reviews = pd.read_csv('reviews_stat.csv', index_col='listing_id')
        df = X.copy().join([calendar, reviews], how='left')
        df.drop(columns='host_id', inplace=True)

        #Convert the column with the percent sign
        df['host_response_rate'] = df['host_response_rate']\
                .str.replace(r'\D+', '', regex=True)\
                .astype('float')

        # Convert datetimes into numbers for further use as a numeric variable
        # This will apply to the host_since column
        df[self.date_features] = df[self.date_features].astype(np.int64)

        result = self.scale_encode.transform(df)
            
        return result

In [4]:
train = pd.read_csv('train.csv',
                    index_col='id',
                    true_values=['t'],
                    false_values=['f'],
                    na_values='none',
                    parse_dates=['host_since'],
                   low_memory=False)
train

name  \
id                                                            
20145777                         Double in a friendly house   
22630537             London flat with great transport links   
27355619  Studio Apartments, 5 mins to Kings Cross Station!   
21011236   Big, Beautiful, Sunny, West Kensington TWIN Room   
24754494                 Dashing 1BR in Bayswater by Sonder   
...                                                     ...   
18151315     Sunny studio with private kitchen and bathroom   
28105053    A comfy double room in modern flat ★ Haggerston   
26342100          Goldsborough House - Apartment London SW8   
13833612                Modern and Luxurious 1 Bedroom Flat   
19640194            SunnyBay - Very Cosy and beautiful home   

                                                    summary  \
id                                                            
20145777  A  double bedroom  in a cozy and modern apartm...   
22630537  Light, airy and modern one bedroom flat 2 mins...   
27355619  Based in high quality student accommodation, o...   
21011236  Big Beautiful, airy, West Kensington TWIN room...   
24754494  At this Sonder, you'll love the chic decor, st...   
...                                                     ...   
18151315  Quiet, sunny studio flat just minutes from the...   
28105053                   It's a double room for 2 guests.   
26342100  Spacious three bedroom flat, fantastic locatio...   
13833612  You’ll love my place because of the ambiance, ...   
19640194  A fantastic two bedroom terraced house situate...   

                                                      space  \
id                                                            
20145777                                                NaN   
22630537                                                NaN   
27355619  What To Expect: * 16/17m2 Studio with modern f...   
21011236  This is a pretty terraced house in a great are...   
24754494  Every booking is instantly confirmed. Every ca...   
...                                                     ...   
18151315  A small and very cosy flat with all necessary ...   
28105053                                                NaN   
26342100                                                NaN   
13833612  This is a beautiful and comfortable one bedroo...   
19640194                                                NaN   

                                                description  \
id                                                            
20145777  A  double bedroom  in a cozy and modern apartm...   
22630537  Light, airy and modern one bedroom flat 2 mins...   
27355619  Based in high quality student accommodation, o...   
21011236  Big Beautiful, airy, West Kensington TWIN room...   
24754494  At this Sonder, you'll love the chic decor, st...   
...                                                     ...   
18151315  Quiet, sunny studio flat just minutes from the...   
28105053  It's a double room for 2 guests. I will be at ...   
26342100  Spacious three bedroom flat, fantastic locatio...   
13833612  You’ll love my place because of the ambiance, ...   
19640194  A fantastic two bedroom terraced house situate...   

         experiences_offered  \
id                             
20145777                 NaN   
22630537                 NaN   
27355619                 NaN   
21011236                 NaN   
24754494                 NaN   
...                      ...   
18151315                 NaN   
28105053                 NaN   
26342100                 NaN   
13833612                 NaN   
19640194                 NaN   

                                      neighborhood_overview  \
id                                                            
20145777                                                NaN   
22630537                                                NaN   
27355619  As one of Central London’s most vibrant and ne...   
21011236  We’re in a fantastic location - c

In [5]:
X = train.drop(columns='price')
y = train.price

In [6]:
X_train, X_val, y_train, y_val = train_test_split(X, y, 
                                                  test_size=0.2,
                                                  random_state=8,
                                                 shuffle=True)

In [7]:
X_val.shape

(7254, 41)

In [8]:
%%time
%%chime
preprocess = Preprocess().fit(X_train)

CPU times: total: 1.05 s
Wall time: 2min 19s


In [9]:
%%time
%%chime
X_train_transf = preprocess.transform(X_train)

CPU times: total: 1.19 s
Wall time: 1min 10s


In [10]:
X_train_transf.shape

(29016, 118)

In [11]:
%%time
%%chime
X_val_transf = preprocess.transform(X_val)

CPU times: total: 703 ms
Wall time: 5.59 s


In [12]:
X_val_transf.shape

(7254, 118)

In [13]:
np.save('X_train_transf', X_train_transf)
np.save('X_val_transf', X_val_transf)
np.save('y_train', y_train)
np.save('y_val', y_val)

In [10]:
X_train_transf = np.load('X_train_transf.npy')
X_val_transf = np.load('X_val_transf.npy')
y_train = np.load('y_train.npy')
y_val = np.load('y_val.npy')

## Предсказание по средней цене

In [16]:
y_val.describe()

count    7254.000000
mean      115.931624
std       175.953193
min         0.000000
25%        45.000000
50%        85.000000
75%       135.000000
max      7716.000000
Name: price, dtype: float64

In [17]:
(y_val - y_val.mean()).abs().mean()

74.18344137864436

## Градиентный бустинг

In [16]:
from sklearn.ensemble import HistGradientBoostingRegressor

In [17]:
params = {'max_iter': [100, 200, 300, 400, 500],
          'max_depth':[10, 20, 50, 70],
          'l2_regularization': [0.0, 0.2, 0.4]
         }

In [18]:
regressor = HistGradientBoostingRegressor(loss='absolute_error',
                                          learning_rate=0.1, 
                                          early_stopping='auto',
                                          scoring='loss',
                                          validation_fraction=0.1,
                                          n_iter_no_change=10,
                                          tol=1e-07, 
                                          random_state=8) 
boost = GridSearchCV(estimator = regressor,
                      param_grid = params,
                      scoring='neg_mean_absolute_error',
                      n_jobs=4, 
                      refit=True,
                      cv=5)

In [19]:
%%time
%%chime
boost.fit(X_train_transf, y_train)

CPU times: total: 15.4 s
Wall time: 12min 13s


In [20]:
boost.best_score_

-33.73235909480901

In [21]:
boost.best_params_

{'l2_regularization': 0.2, 'max_depth': 20, 'max_iter': 400}

Сужаю сеть

In [22]:
params = {'max_iter': [350, 400, 450],
          'max_depth':[15, 20, 30, 40],
          'l2_regularization': [0.1, 0.2, 0.3]
         }

In [23]:
regressor = HistGradientBoostingRegressor(loss='absolute_error',
                                          learning_rate=0.1, 
                                          early_stopping='auto',
                                          scoring='loss',
                                          validation_fraction=0.1,
                                          n_iter_no_change=10,
                                          tol=1e-07, 
                                          random_state=8) 
boost = GridSearchCV(estimator = regressor,
                      param_grid = params,
                      scoring='neg_mean_absolute_error',
                      n_jobs=4, 
                      refit=True,
                      cv=5)

In [24]:
%%time
%%chime
boost.fit(X_train_transf, y_train)

CPU times: total: 15 s
Wall time: 8min 27s


In [25]:
boost.best_score_

-33.73235909480901

In [26]:
boost.best_params_

{'l2_regularization': 0.2, 'max_depth': 20, 'max_iter': 350}

In [28]:
cv_results = pd.DataFrame(boost.cv_results_); cv_results

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       10.713933      1.339673         0.433677        0.063380   
1       10.547916      1.436143         0.481555        0.094631   
2       10.562870      1.247340         0.452992        0.038339   
3        9.728473      1.870455         0.415648        0.092496   
4        9.622375      1.940959         0.425022        0.105704   
5        9.848118      2.009232         0.403145        0.082329   
6        9.838011      1.988094         0.421897        0.086155   
7        9.756324      1.838512         0.409396        0.088061   
8        9.716130      1.840994         0.428147        0.114229   
9        9.702237      1.931150         0.414136        0.078169   
10       9.893242      2.016675         0.393771        0.080529   
11       9.809886      1.960310         0.434397        0.101461   
12       9.786777      3.493032         0.437523        0.162389   
13      10.272407      3.994608         0.462525        0.230627   
14      10.474859      4.165367         0.412522        0.184516   
15      11.404532      1.587327         0.487523        0.073559   
16      11.283299      1.573594         0.488014        0.085283   
17      11.256835      1.581889         0.493775        0.094067   
18      11.293022      1.578046         0.479759        0.066522   
19      11.300789      1.477756         0.496902        0.111108   
20      11.321066      1.420267         0.471899        0.092390   
21      11.315747      1.441921         0.471900        0.047805   
22      11.275536      1.544799         0.471898        0.057964   
23      11.265023      1.744871         0.481274        0.054489   
24      11.049997      2.258937         0.453150        0.107354   
25      11.293763      2.516333         0.465645        0.115418   
26      11.253401      2.502088         0.478150        0.109870   
27      11.053390      2.346356         0.462525        0.133176   
28      10.799126      2.357911         0.462525        0.122879   
29      10.936633      2.271988         0.465648        0.136008   
30      10.954303      2.265357         0.462523        0.114656   
31      10.923051      2.303249         0.443774        0.102035   
32      10.975301      2.404021         0.443773        0.112938   
33      10.931548      2.596322         0.453148        0.088943   
34      10.941193      2.489037         0.450269        0.109378   
35       9.971384      2.314604         0.335208        0.136812   

   param_l2_regularization param_max_depth param_max_iter  \
0                      0.1              15            350   
1                      0.1              15            400   
2                      0.1              15            450   
3                      0.1              20            350   
4                      0.1              20            400   
5                      0.1              20            450   
6                      0.1              30            350   
7                      0.1              30            400   
8                      0.1              30            450   
9                      0.1              40            350   
10                     0.1              40            400   
11                     0.1              40            450   
12                     0.2              15            350   
13                     0.2              15            400   
14                     0.2              15            450   
15                     0.2              20            350   
16                     0.2              20            400   
17                     0.2              20            450   
18                     0.2              30            350   
19                     0.2              30            400   
20                     0.2              30            450   
21                     0.2              40            350   
22                     0.2              40            400   
23                     

In [29]:
cv_results.rank_test_score.unique()

array([10, 28, 27, 26, 25,  1, 24, 22, 13])

In [30]:
cv_results.query('rank_test_score == 1')

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
15      11.404532      1.587327         0.487523        0.073559   
16      11.283299      1.573594         0.488014        0.085283   
17      11.256835      1.581889         0.493775        0.094067   
18      11.293022      1.578046         0.479759        0.066522   
19      11.300789      1.477756         0.496902        0.111108   
20      11.321066      1.420267         0.471899        0.092390   
21      11.315747      1.441921         0.471900        0.047805   
22      11.275536      1.544799         0.471898        0.057964   
23      11.265023      1.744871         0.481274        0.054489   

   param_l2_regularization param_max_depth param_max_iter  \
15                     0.2              20            350   
16                     0.2              20            400   
17                     0.2              20            450   
18                     0.2              30            350   
19                     0.2              30            400   
20                     0.2              30            450   
21                     0.2              40            350   
22                     0.2              40            400   
23                     0.2              40            450   

                                               params  split0_test_score  \
15  {'l2_regularization': 0.2, 'max_depth': 20, 'm...         -31.324618   
16  {'l2_regularization': 0.2, 'max_depth': 20, 'm...         -31.324618   
17  {'l2_regularization': 0.2, 'max_depth': 20, 'm...         -31.324618   
18  {'l2_regularization': 0.2, 'max_depth': 30, 'm...         -31.324618   
19  {'l2_regularization': 0.2, 'max_depth': 30, 'm...         -31.324618   
20  {'l2_regularization': 0.2, 'max_depth': 30, 'm...         -31.324618   
21  {'l2_regularization': 0.2, 'max_depth': 40, 'm...         -31.324618   
22  {'l2_regularization': 0.2, 'max_depth': 40, 'm...         -31.324618   
23  {'l2_regularization': 0.2, 'max_depth': 40, 'm...         -31.324618   

    split1_test_score  split2_test_score  split3_test_score  \
15         -33.217148         -33.015609         -37.097249   
16         -33.217148         -33.015609         -37.097249   
17         -33.217148         -33.015609         -37.097249   
18         -33.217148         -33.015609         -37.097249   
19         -33.217148         -33.015609         -37.097249   
20         -33.217148         -33.015609         -37.097249   
21         -33.217148         -33.015609         -37.097249   
22         -33.217148         -33.015609         -37.097249   
23         -33.217148         -33.015609         -37.097249   

    split4_test_score  mean_test_score  std_test_score  rank_test_score  
15         -34.007171       -33.732359        1.896017                1  
16         -34.007171       -33.732359        1.896017                1  
17         -34.007171       -33.732359        1.896017                1  
18         -34.007171       -33.732359        1.896017                1  
19         -34.007171       -33.732359        1.896017                1  
20         -34.007171       -33.732359        1.896017                1  
21         -34.007171       -33.732359        1.896017                1  
22         -34.007171       -33.732359        1.896017                1  
23         -34.007171       -33.732359        1.896017                1

Результаты немного лучше, чем у того же метода с предварительной подстановкой пропущенных значений: 33.73 против 34.04, но стандартное отклонение кроссвалидации 1.90 говорит о том, что изменение не существенное. Тем не менее буду испльзовать этот вариант, т.к. он проще.

# Окончательная модель и прогноз

In [37]:
from joblib import dump, load

In [38]:
best_estimator = HistGradientBoostingRegressor(loss='absolute_error',
                                              l2_regularization=0.2,
                                              max_depth=20,
                                              max_iter =350,
                                              learning_rate=0.1, 
                                              early_stopping='auto',
                                              scoring='loss',
                                              validation_fraction=0.1,
                                              n_iter_no_change=10,
                                              tol=1e-07, 
                                              random_state=8)

Окончательный конвеер, включающий подготовку данных и регрессионную модель

In [39]:
pipe = Pipeline([
    ('prep', Preprocess()),
    ('regress', best_estimator)
])

In [40]:
%%time
%%chime
pipe.fit(X_train, y_train)

CPU times: total: 16 s
Wall time: 3min 30s


In [41]:
y_pred_val = pipe.predict(X_val)

In [42]:
mean_absolute_error(y_val, y_pred_val)

33.678846077603616

In [43]:
dump(pipe, 'best_model.joblib')

['best_model.joblib']

### Прогноз на тестовой выборке

In [44]:
X_test = pd.read_csv('test.csv',
                    index_col='id',
                    true_values=['t'],
                    false_values=['f'],
                    na_values='none',
                    parse_dates=['host_since'],
                   low_memory=False)
X_test

name  \
id                                                            
24366684  Luxurious 2 Bed by Kings Road, Chelsea with Ga...   
20329256                  Fulham Chelsea Great Studio! HR1a   
27087563  Smart Spacious Double Room with Kitchenette & ...   
21995708    Beautiful 1 bedroom garden flat with lovely cat   
19372467                  The Bolt Hole [440a, Fulham Road]   
...                                                     ...   
19819730               Discover Shoreditch & Hackney "Pear"   
27683820                                      56 Haldane Rd   
21011115               A bright 2 bedroom flat in Islington   
29033913   Amazing Apartment Near Hyde Park & Oxford Street   
27874952             Central London home near Clissold Park   

                                                    summary  \
id                                                            
24366684  This stunning newly-refurbished 2 bedroom Chel...   
20329256  My place is close to Chelsea Football Ground. ...   
27087563  A spacious double room with your own private b...   
21995708  Beautifully decorated and peaceful 1 bedroom f...   
19372467  The Bolt Hole is a delightful 2 bedroom house ...   
...                                                     ...   
19819730  **** Instagram @host.tay **** The Old Nags Hea...   
27683820  Hello! I am the owner Erin J. Buchholz. Welcom...   
21011115  A sunny family home with a garden, 2 double be...   
29033913                                                NaN   
27874952  Victorian house with a private bedroom & bathr...   

                                                      space  \
id                                                            
24366684  Amenities:  - 24/7 guest support - Professiona...   
20329256  cosy self contained studio - close to public t...   
27087563  A bright sunny double room with a private bath...   
21995708                                                NaN   
19372467  An unusual house, built in the 1960's, The Bol...   
...                                                     ...   
19819730  The apartment has recently been renovated so e...   
27683820                                                NaN   
21011115  Our flat is homey, quiet and has a lovely priv...   
29033913                                                NaN   
27874952                                                NaN   

                                                description  \
id                                                            
24366684  This stunning newly-refurbished 2 bedroom Chel...   
20329256  My place is close to Chelsea Football Ground. ...   
27087563  A spacious double room with your own private b...   
21995708  Beautifully decorated and peaceful 1 bedroom f...   
19372467  The Bolt Hole is a delightful 2 bedroom house ...   
...                                                     ...   
19819730  **** Instagram @host.tay **** The Old Nags Hea...   
27683820  Hello! I am the owner Erin J. Buchholz. Welcom...   
21011115  A sunny family home with a garden, 2 double be...   
29033913                                                NaN   
27874952  Victorian house with a private bedroom & bathr...   

         experiences_offered  \
id                             
24366684                 NaN   
20329256                 NaN   
27087563                 NaN   
21995708                 NaN   
19372467                 NaN   
...                      ...   
19819730                 NaN   
27683820                 NaN   
21011115                 NaN   
29033913                 NaN   
27874952                 NaN   

                                      neighborhood_overview  \
id                                                            
24366684  Two long high streets, King's Road and Fulham ...   
20329256  Fulham is a cool, vibrant area with great shop...   
27087563  The house located in zone 2, in a quiet reside...   
21995708  The house is at the heart of Lond

In [47]:
%%time
%%chime
y_pred_test = pipe.predict(X_test)

CPU times: total: 1.47 s
Wall time: 21.1 s


In [48]:
y_pred_test = pd.DataFrame(y_pred_test, columns=['price'], index=X_test.index)
y_pred_test

price
id                  
24366684  161.703510
20329256   76.553143
27087563   45.162014
21995708   60.841133
19372467  125.670140
...              ...
19819730   41.977211
27683820  120.984147
21011115  100.427615
29033913  173.214637
27874952   44.021567

[15545 rows x 1 columns]

In [49]:
y_pred_test.to_csv('submission.csv')